File: apps/mobile/lib/services/coach/coach_models.dart
/// Coach shared models — Sprint S34.
///
/// Data classes for compliance validation, coach context,
/// and hallucination detection.
library;

/// Type of coach output component, each with length constraints.
enum ComponentType {
  greeting,          // max 30 words
  scoreSummary,      // max 80 words
  tip,               // max 120 words
  chiffreChoc,       // max 100 words
  scenario,          // max 150 words
  enrichmentGuide,   // max 150 words — data block conversational guide
  chatSystem,        // max 300 words — main chat system prompt (S51)
  chatSafeMode,      // max 200 words — debt/stress mode (S51)
  chatFollowUp,      // max 250 words — multi-turn follow-up (S51)
  chatSimulation,    // max 250 words — simulation scenario (S51)
  chatSenior,        // max 200 words — adapted for 60+ users (S51)
  general,           // max 200 words
}

/// Word limits per component type.
const Map<ComponentType, int> componentWordLimits = {
  ComponentType.greeting: 30,
  ComponentType.scoreSummary: 80,
  ComponentType.tip: 120,
  ComponentType.chiffreChoc: 100,
  ComponentType.scenario: 150,
  ComponentType.enrichmentGuide: 150,
  ComponentType.chatSystem: 300,
  ComponentType.chatSafeMode: 200,
  ComponentType.chatFollowUp: 250,
  ComponentType.chatSimulation: 250,
  ComponentType.chatSenior: 200,
  ComponentType.general: 200,
};

/// Result of compliance validation on LLM output.
class ComplianceResult {
  final bool isCompliant;
  final String sanitizedText;
  final List<String> violations;
  final bool useFallback;

  const ComplianceResult({
    required this.isCompliant,
    required this.sanitizedText,
    this.violations = const [],
    this.useFallback = false,
  });
}

/// Context passed to compliance guard and narrative generation.
///
/// Contains financial_core outputs — NEVER raw amounts.
/// All values are pre-computed by financial_core calculators.
///
/// Extended in S35 with archetype, behavioral, and temporal fields
/// to support the Coach Narrative Service.
class CoachContext {
  final String firstName;
  final String archetype;
  final int age;
  final String canton;
  // Financial state (aggregated, never raw)
  final double friTotal;
  final double friDelta;
  final String primaryFocus;
  final double replacementRatio;
  final double monthsLiquidity;
  final double taxSavingPotential;
  final double confidenceScore;
  // Temporal
  final int daysSinceLastVisit;
  final String fiscalSeason;
  final String upcomingEvent;
  // Behavioral
  final int checkInStreak;
  final String lastMilestone;
  // Known numerical values for hallucination detection
  final Map<String, double> knownValues;
  // Data reliability by field: 'certified', 'userInput', or 'estimated'
  // e.g. {'avoirLpp': 'certified', 'patrimoine': 'estimated'}
  final Map<String, String> dataReliability;

  const CoachContext({
    this.firstName = 'utilisateur',
    this.archetype = 'swiss_native',
    this.age = 30,
    this.canton = 'ZH',
    this.friTotal = 0.0,
    this.friDelta = 0.0,
    this.primaryFocus = '',
    this.replacementRatio = 0.0,
    this.monthsLiquidity = 0.0,
    this.taxSavingPotential = 0.0,
    this.confidenceScore = 0.0,
    this.daysSinceLastVisit = 0,
    this.fiscalSeason = '',
    this.upcomingEvent = '',
    this.checkInStreak = 0,
    this.lastMilestone = '',
    this.knownValues = const {},
    this.dataReliability = const {},
  });

  CoachContext copyWith({String? fiscalSeason}) {
    return CoachContext(
      firstName: firstName,
      archetype: archetype,
      age: age,
      canton: canton,
      friTotal: friTotal,
      friDelta: friDelta,
      primaryFocus: primaryFocus,
      replacementRatio: replacementRatio,
      monthsLiquidity: monthsLiquidity,
      taxSavingPotential: taxSavingPotential,
      confidenceScore: confidenceScore,
      daysSinceLastVisit: daysSinceLastVisit,
      fiscalSeason: fiscalSeason ?? this.fiscalSeason,
      upcomingEvent: upcomingEvent,
      checkInStreak: checkInStreak,
      lastMilestone: lastMilestone,
      knownValues: knownValues,
      dataReliability: dataReliability,
    );
  }
}

/// A number found in LLM output that doesn't match known values.
class HallucinatedNumber {
  final String foundText;
  final double foundValue;
  final String closestKey;
  final double closestValue;
  final double deviationPct;

  const HallucinatedNumber({
    required this.foundText,
    required this.foundValue,
    required this.closestKey,
    required this.closestValue,
    required this.deviationPct,
  });
}

File: services/backend/app/services/coach/coach_models.py
"""
Coach shared models — Sprint S34 + S35.

Data classes for compliance validation, coach context, hallucination detection,
and narrative generation results.
"""

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Optional


class ComponentType(str, Enum):
    """Type of coach output component, each with length constraints."""
    greeting = "greeting"                # max 30 words
    score_summary = "score_summary"      # max 80 words
    tip = "tip"                          # max 120 words
    premier_eclairage = "premier_eclairage"        # max 100 words
    scenario = "scenario"                # max 150 words
    enrichment_guide = "enrichmentGuide" # max 150 words — data block conversational guide
    general = "general"                  # max 200 words


# Word limits per component type
COMPONENT_WORD_LIMITS = {
    ComponentType.greeting: 30,
    ComponentType.score_summary: 80,
    ComponentType.tip: 120,
    ComponentType.premier_eclairage: 100,
    ComponentType.scenario: 150,
    ComponentType.enrichment_guide: 150,
    ComponentType.general: 200,
}


@dataclass
class ComplianceResult:
    """Result of compliance validation on LLM output."""
    is_compliant: bool
    sanitized_text: str              # Cleaned version (if salvageable)
    violations: List[str] = field(default_factory=list)
    use_fallback: bool = False       # If True, discard LLM output entirely


@dataclass
class CoachContext:
    """Context passed to compliance guard and narrative generation.

    Contains financial_core outputs — NEVER raw amounts.
    All values are pre-computed by financial_core calculators.

    Extended in S35 with archetype, behavioral, and temporal fields
    to support the Coach Narrative Service.
    """
    first_name: str = "utilisateur"
    # B4 fix (2026-05-08) : defaults set to empty/zero so callers that
    # don't provide these fields don't leak fake « known facts » into the
    # LLM context. Consumers (claude_coach_service.py:793-798) gate prompt
    # lines on truthy values, so an empty canton/archetype/age means the
    # LLM is told « unknown — ask the user », not « VD / swiss_native /
    # 30 ans » (false confidence reported in TestFlight v2.12.1+3).
    archetype: str = ""
    age: int = 0
    canton: str = ""
    # Financial state (aggregated, never raw)
    fri_total: float = 0.0
    fri_delta: float = 0.0
    primary_focus: str = ""
    replacement_ratio: float = 0.0
    months_liquidity: float = 0.0
    tax_saving_potential: float = 0.0
    confidence_score: float = 0.0
    # Temporal
    days_since_last_visit: int = 0
    fiscal_season: str = ""    # "3a_deadline", "tax_declaration", ""
    upcoming_event: str = ""
    # Behavioral
    check_in_streak: int = 0
    last_milestone: str = ""
    # Check-in: list of planned contributions for sequential check-in conversation
    # Each item: {'id': '3a_julien', 'label': '3e pilier (Julien)', 'amount': 500.0}
    planned_contributions: list = field(default_factory=list)
    # Known numerical values for hallucination detection
    known_values: dict = field(default_factory=dict)
    # Structured mobile data-spine packet. Already privacy-scoped by mobile
    # CoachContextPacketService and sanitized again at the API boundary.
    coach_context_packet: Dict[str, Any] = field(default_factory=dict)
    # Onboarding intent (e.g., 'firstJob', 'intentChip3a', etc.)
    intent: str = ""
    # Safe Mode signal: consumer debt stress or emergency-fund shortfall.
    # When True, the system prompt injects _SAFE_MODE_PROTOCOL to block
    # all 3a/LPP optimisation advice (RULES.md §2).
    has_debt: bool = False
    # firstJob P0 #1114 — contexte MoneyTruthReceipt résolu server-side pour
    # ce tour (owner-scoped). Forme = sortie de `resolve_receipt_context`
    # ({"status": "resolved"|"pending"|"not_found", ...}). None = pas de
    # receipt sur ce tour -> chemin coach classique STRICTEMENT inchangé.
    # Non whitelisté par _build_coach_context_from_profile (dict, pas un
    # scalaire) : rattaché explicitement pour atteindre le prompt via
    # _build_context_section (sinon façade sans câblage : valeur droppée).
    money_truth_receipt: Optional[Dict[str, Any]] = None


@dataclass
class CoachNarrativeResult:
    """Result of the Coach Narrative Service — 4 independent components.

    Each component is generated independently and validated through
    ComplianceGuard before reaching the user.

    Sources:
        - LSFin art. 3 (quality of financial information)
        - LPD art. 6 (data processing principles)
    """
    greeting: str
    score_summary: str
    tip_narrative: str
    premier_eclairage_reframe: str
    used_fallback: Dict[str, bool] = field(default_factory=dict)
    disclaimer: str = ""
    sources: List[str] = field(default_factory=list)


@dataclass
class HallucinatedNumber:
    """A number found in LLM output that doesn't match known values."""
    found_text: str          # Original text fragment
    found_value: float       # Parsed numeric value
    closest_key: str         # Which known value it was compared against
    closest_value: float     # The expected value
    deviation_pct: float     # How much it deviates (as percentage)
